### PR TITLE
fix(tcas): TCAS critical hotfixes and improvements

### LINKED ISSUE #5089: Altitude Reporting doesn't shut the TCAS to STBY when OFF
### Aircraft Version

Development

### Build info

```json
{
  "built": "2021-06-10T09:49:19+00:00",
  "ref": "master",
  "sha": "1578f4fa852800ef1d7c46ad6460f21bf92a2c3a",
  "actor": "MisterChocker",
  "event_name": "manual"
}
```

### Describe the bug

When ALT RPTG is OFF, nothing happens to TCAS

### Expected behavior

When ALT RPTG is OFF, TCAS is expected to become STBY (Recommended procedure when operating above 1050hPa)

### Steps to reproduce

1. Turn TCAS to TA/RA
2. Turn ALT RPTG to OFF
3. TCAS is expected to become STBY but it doesn't.

### References (optional)

image

### Additional info (optional)

_No response_

### Discord Username (optional)

Illuminati

### PR DESCRIPTION
**Pull request by 2hwk**
Fixes #5089

## Summary of Changes

* Fixed issue with ATC XPDR knob having missing tooltip.
* Fixed issue with ATC RPTG not triggering TCAS STBY alert and not disabling TCAS reporting
* Fixed plane spawn state on runway: ATC RPTG should be ON not OFF
* Linked green TCAS STBY alert behaviour with TCAS sensitivity level (thanks Beheh)
* Improved TCAS sensitivity level calculation logic
* Minor refactor/streamlining of TCAS computer logic
* Added more documentation/code comments

## References

See #5089 and #3378

## Additional context

Discord username (if different from GitHub): 2Cas#1022

## Testing instructions

* Spawn in C&D + normal engine start procedure (a major airport would be good for convenience for traffic)
* Turn on EFB, go to settings > Sim Options, TCAS debug option should have been removed and be no longer present. 
* Ensure that ATC XPDR selection knob has a proper tool-tip description.
* You can either inject traffic in via ATC network, a traffic add-on or use MSFS live traffic (not AI Offline traffic). 
* Take note of TCAS STBY green alert on the right side of the Upper ECAM. This should now correctly toggle on when ALT RPTG and TCAS STBY switches are set to OFF and STBY. (TA ONLY should not appear on the ND when green TCAS STBY is displayed and TCAS is inactive)
* Takeoff with either ALT RPTG switch at OFF position, or the TCAS switch in the STBY position (not TA/RA or TA). After wheels up takeoff, the alert should turn into an amber caution (NAV TCAS STBY and TCAS STBY in amber) on the upper ECAM.
* Switch ALT RPTG to ON and TCAS to TA/RA. The caution should disappear and intruder/traffic/TCAS functionality should be restored on the ND.
* Use the dev mode Windows > Behavior debug window to modify the A32NX_SWITCH_ATC_ALT LVar to manipulate the ALT RPTG knob, see:

https://user-images.githubusercontent.com/15316958/155104546-55efd558-1c84-4c99-bd41-a87280c7aa50.mp4

* TA/RAs should be issued where appropriate, traffic should be displayed as per defined envelops depending on reporting mode (defined range: 80 nm ahead, 30nm side to side, 20 nm behind, with mode set to ABV = + 9900 ft/-2700 ft, BLW = -9900 ft/+2700ft, ALL +-2700ft. Note that MSFS Live Traffic can be buggy and show contacts that are invisible/not present/impossible, it is better to rely on traffic addons/VATSIM for this.)

* Spawn in mid-air and on the runway, ALT RPTG switch should be ON. Ensure that the TCAS STBY memo is not active on the Upper ECAM (TCAS should be functioning as normal). 

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/Climb.flt
Comment (tracernz) [inline flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/Climb.flt:148]: Please keep this `A32NX_TRANSPONDER_ALT_RPTG` and fix the changes that the tcas PR made.

**Review by tracernz — CHANGES_REQUESTED** [review]

File: flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/model/A320_NEO_INTERIOR.xml
Comment (tracernz) [inline flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/model/A320_NEO_INTERIOR.xml:4315]: What is this one for? KNOB_ATC_1 is already covered by the main transponder template.

**Review by tracernz — APPROVED** [review]

**Comment by AndLindemann** [comment]
This PR (as it is in experimental) has removed the A32NX_SWITCH_ATC_ALT setting we could use to bind to controls for flipping the ALT RPTG knob in the cockpit. Is this intended?
Setting L:A32NX_TRANSPONDER_ALT_RPTG to 0 does not properly flip the knob back into the Off position, so cannot currently be used in the same way for me.

**Comment by 2hwk** [comment]
> This PR (as it is in experimental) has removed the A32NX_SWITCH_ATC_ALT setting we could use to bind to controls for flipping the ALT RPTG knob in the cockpit. Is this intended? Setting L:A32NX_TRANSPONDER_ALT_RPTG to 0 does not properly flip the knob back into the Off position, so cannot currently be used in the same way for me. It does switch off the actual altitude reporting (and TCAS), but the cockpit knob remains in the "on" position, when the change is done through an externaly bound control (like spad.next).

I'll put this on hold for now then, until the best way can be found to revert this and break as little as possible.

**Comment by AndLindemann** [comment]
> > This PR (as it is in experimental) has removed the A32NX_SWITCH_ATC_ALT setting we could use to bind to controls for flipping the ALT RPTG knob in the cockpit. Is this intended? Setting L:A32NX_TRANSPONDER_ALT_RPTG to 0 does not properly flip the knob back into the Off position, so cannot currently be used in the same way for me. It does switch off the actual altitude reporting (and TCAS), but the cockpit knob remains in the "on" position, when the change is done through an externaly bound control (like spad.next).
> 
> I'll put this on hold for now then, until the best way can be found to revert this and break as little as possible.

I'm not an expert, but could you not just tie the cockpit knob position (behavior template?) directly to the value of the A32NX_TRANSPONDER_ALT_RPTG  local var, in the same way as it before was tied into the A32NX_SWITCH_ATC_ALT? Flight Deck API docs would need an update then as they specifically refer to the A32NX_SWITCH_ATC_ALT for flipping the knob.
Or otherwise, what reasons are there for removing A32NX_SWITCH_ATC_ALT in the first place, can it not just be kept for the knob state/action?

**Review by bouveng — APPROVED** [review]

**Comment by 2hwk** [comment]
The solution:
Consolidating the two SimVars into one will still occur, but A32NX_TRANSPONDER_ALT_RPTG will be renamed to A32NX_SWITCH_ATC_ALT. 

Note that LocalVars and SimVars which are not documented by the list in .github/docs/a320-simvars.md can and will be subject to change, I'm just allowing for this, as it's been brought up before merge and thus it's simple to resolve. This may not always be the case. I've added the var to the list so that in future this will be more locked in place.

File: src/behavior/src/A32NX_Interior_ATC.xml
Comment (AndLindemann) [inline src/behavior/src/A32NX_Interior_ATC.xml:227]: It's not about renaming the variable. It's that A32NX_SWITCH_ALT_RPTG did not exhibit the same behavior than what A32NX_SWITCH_ATL_ALT did previously, in that changing the var does not make the cockpit switch reflect the correct position.
I wonder if removing these two lines is what actually caused the cockpit switch glitch - but I'm really not familiar with how knob animation ties into LVARS, so it might just be me.

Before this change, setting A32NX_SWITCH_ATC_ALT to 0 via i.e. spad.next caused the cockpit switch to flip into the "off" position.
After your changes, setting A32NX_TRANSPONDER_ALT_RPTG to 0 did not cause the cockpit switch to flip into "off", it would remain in "on". That's all I'm saying - it's no problem if the variable name changes, as long as it still properly interacts with the cockpit knob.
Not sure if it's clear this way.
I haven't tested your latest change yet, so maybe that resolves this. Once this is available on exp, I can give it a shot.

Comment (2hwk) [reply src/behavior/src/A32NX_Interior_ATC.xml:227]: I didn't actually push the correct change to the repo, it's been pushed now.

File: src/behavior/src/A32NX_Interior_ATC.xml
Comment (2hwk) [inline src/behavior/src/A32NX_Interior_ATC.xml:230]: Knob animation should now follow the value of L:A32NX_SWITCH_ATC_ALT

File: src/behavior/src/A32NX_Interior_ATC.xml
Comment (AndLindemann) [reply src/behavior/src/A32NX_Interior_ATC.xml:227]: Thanks and sorry for the extra work. I'll give it a shot once it is available on experimental or development.

**Comment by ferryvanaesch** [comment]
Not entirely sure if it's related to this PR or not, but lately when I change mode from Standby to Auto, Mode C (in vPilot, VATSIM) switches on, even when I'm still shut down at the gate. This used to switch on during pushback and engine start. Is that intentional?

**Comment by AndLindemann** [comment]
> Not entirely sure if it's related to this PR or not, but lately when I change mode from Standby to Auto, Mode C (in vPilot, VATSIM) switches on, even when I'm still shut down at the gate. This used to switch on during pushback and engine start. Is that intentional?

I think that is from a different PR https://github.com/flybywiresim/a32nx/pull/6697
What you see is the transponder being in mode S (ground mode), which is seen as mode C by vPilot, but that seems to be intentional. Not related to this PR.

**Comment by ferryvanaesch** [comment]
> > Not entirely sure if it's related to this PR or not, but lately when I change mode from Standby to Auto, Mode C (in vPilot, VATSIM) switches on, even when I'm still shut down at the gate. This used to switch on during pushback and engine start. Is that intentional?
> 
> I think that is from a different PR #6697 What you see is the transponder being in mode S (ground mode), which is seen as mode C by vPilot, but that seems to be intentional. Not related to this PR. While in ground mode, the transponder will not transmit altitude, even when ALT RPTG is on. It will automatically switch to mode C during take-off (when in Auto position) and will then transmit altitude if ALT RPTG is set to on.

Totally missed that PR. Thanks!

**Comment by tracernz** [comment]
> Note that LocalVars and SimVars which are not documented by the list in .github/docs/a320-simvars.md can and will be subject to change, I'm just allowing for this, as it's been brought up before merge and thus it's simple to resolve. This may not always be the case. I've added the var to the list so that in future this will be more locked in place.

The ones documented in there are also subject to change. It's just the nature of an aircraft this early in it's development cycle. 😁

**Comment by mico975** [comment]
Quality Assurance Tester Report

Discord          : mico#3145
Object of testing: #6766 
Tier of Testing  : 1/
Date             : 23/02/2022

Testing Process: 

Turn on EFB, go to settings > Sim Options, TCAS debug option should have been removed and be no longer present. - **OK**
Ensure that ATC XPDR selection knob has a proper tool-tip description. - **NOT OK**, it's inverted, see screenshot below
Take note of TCAS STBY green alert on the right side of the Upper ECAM. This should now correctly toggle on when ALT RPTG and TCAS STBY switches are set to OFF and STBY. (TA ONLY should not appear on the ND when green TCAS STBY is displayed and TCAS is inactive) - **OK**
Takeoff with either ALT RPTG switch at OFF position, or the TCAS switch in the STBY position (not TA/RA or TA). After wheels up takeoff, the alert should turn into an amber caution (NAV TCAS STBY and TCAS STBY in amber) on the upper ECAM. - **OK**
Switch ALT RPTG to ON and TCAS to TA/RA. The caution should disappear and intruder/traffic/TCAS functionality should be restored on the ND. - **OK**
Use the dev mode Windows > Behavior debug window to modify the A32NX_SWITCH_ATC_ALT LVar to manipulate the ALT RPTG knob - **OK**
TA/RAs should be issued where appropriate, traffic should be displayed as per defined envelops depending on reporting mode (defined range: 80 nm ahead, 30nm side to side, 20 nm behind, with mode set to ABV = + 9900 ft/-2700 ft, BLW = -9900 ft/+2700ft, ALL +-2700ft. Note that MSFS Live Traffic can be buggy and show contacts that are invisible/not present/impossible, it is better to rely on traffic addons/VATSIM for this.) - **OK**
Spawn in mid-air and on the runway, ALT RPTG switch should be ON. Ensure that the TCAS STBY memo is not active on the Upper ECAM (TCAS should be functioning as normal). - **OK**

Negatives:       
ALT RPTG tooltip is inverted. Also transponder switch has "SET TRANSPONDER TO TRANSPONER" when it should probably be "SET TRANSPONDER TO TRANSPONDER 2" and vice versa.

Testing Results: 
Not Passed

Conclusions: 
Considered passed when tooltips get corrected. It's merged though 🤣 

Media:       
image

**Comment by 2hwk** [comment]
> > Note that LocalVars and SimVars which are not documented by the list in .github/docs/a320-simvars.md can and will be subject to change, I'm just allowing for this, as it's been brought up before merge and thus it's simple to resolve. This may not always be the case. I've added the var to the list so that in future this will be more locked in place.
> 
> The ones documented in there are also subject to change. It's just the nature of an aircraft this early in it's development cycle. 😁

Doubly so for ones that aren't documented! :P

File: src/behavior/src/A32NX_Interior_ATC.xml
Comment (2hwk) [inline src/behavior/src/A32NX_Interior_ATC.xml:237]: A PR to swap these around would be appreciated but its not vital

**Pull request merged**
